File: src/compression.go
package main

import "bytes"

func RLECompress(buf []byte) []byte {
    var outBuf bytes.Buffer
    prevByte := buf[0]
    count := byte(1)

    for _, b := range buf[1:] {
        if b == prevByte && count < 255 {
            count++
        } else {
            outBuf.WriteByte(count)
            outBuf.WriteByte(prevByte)
            prevByte = b
            count = 1
        }
    }

    outBuf.WriteByte(count)
    outBuf.WriteByte(prevByte)

    return outBuf.Bytes()
}

func RLEDecompress(buf []byte) []byte {
    var outBuf []byte

    for i := 0; i < len(buf) - 1; i+=2 {
        count := int(buf[i])
        b     := buf[i + 1]
        for range count {
            outBuf = append(outBuf, b)
        }
    }

    return outBuf
}

File: decryptor/compression.c
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

uint8_t* RLEDecompress(const uint8_t* buf, size_t bufLen, size_t* outLen) {
    if (bufLen & 1) return NULL;

    size_t capacity = 0;
    for (size_t i = 0; i + 1 < bufLen; i += 2) {
        capacity += buf[i];
    }

    uint8_t* outBuf = malloc(capacity);
    if (!outBuf) return NULL;

    size_t pos = 0;
    for (size_t i = 0; i + 1 < bufLen; i += 2) {
        uint8_t count = buf[i];
        uint8_t val   = buf[i+1];
        memset(outBuf + pos, val, count);
        pos += count;
    }

    *outLen = pos;
    return outBuf;
}
